diesel_derives/src/sql_function.rs

```diff
@@ -0,0 +1,285 @@
+
+    let mut tokens = quote! {
+        use diesel::{self, QueryResult};
+        use diesel::expression::{AsExpression, Expression, SelectableExpression, AppearsOnTable};
+        use diesel::query_builder::{QueryFragment, AstPass};
+        use diesel::sql_types::*;
+        use super::*;
+
+        #[derive(Debug, Clone, Copy, QueryId, DieselNumericOps)]
+        pub struct #fn_name #ty_generics {
+            #(pub(in super) #args,)*
+            #(pub(in super) #type_args: ::std::marker::PhantomData<#type_args2>,)*
+        }
+
+        pub type HelperType #ty_generics = #fn_name <
+            #(#type_args,)*
+            #(<#arg_name as AsExpression<#arg_type>>::Expression,)*
+        >;
+
+        impl #impl_generics Expression for #fn_name #ty_generics
+        #where_clause
+            (#(#arg_name),*): Expression,
+        {
+            type SqlType = #return_type;
+        }
+
+        // __DieselInternal is what we call QS normally
+        impl #impl_generics_internal SelectableExpression<__DieselInternal>
+            for #fn_name #ty_generics
+        #where_clause
+            #(#arg_name: SelectableExpression<__DieselInternal>,)*
+            Self: AppearsOnTable<__DieselInternal>,
+        {
+        }
+
+        // __DieselInternal is what we call QS normally
+        impl #impl_generics_internal AppearsOnTable<__DieselInternal>
+            for #fn_name #ty_generics
+        #where_clause
+            #(#arg_name: AppearsOnTable<__DieselInternal>,)*
+            Self: Expression,
+        {
+        }
+
+        // __DieselInternal is what we call DB normally
+        impl #impl_generics_internal QueryFragment<__DieselInternal>
+            for #fn_name #ty_generics
+        where
+            __DieselInternal: diesel::backend::Backend,
+            for<'a> (#(&'a #arg_name),*): QueryFragment<__DieselInternal>,
+        {
+            fn walk_ast(&self, mut out: AstPass<__DieselInternal>) -> QueryResult<()> {
+                out.push_sql(concat!(#sql_name, "("));
+                (#(&self.#arg_name,)*).walk_ast(out.reborrow())?;
+                out.push_sql(")");
+                Ok(())
+            }
+        }
+    };
+
+    if !is_aggregate {
+        tokens = quote! {
+            #tokens
+
+            impl #impl_generics diesel::expression::NonAggregate
+                for #fn_name #ty_generics
+            #where_clause
+                #(#arg_name: diesel::expression::NonAggregate,)*
+                Self: Expression,
+            {
+            }
+        };
+
+        if cfg!(feature = "sqlite") && type_args.is_empty() {
+            tokens = quote! {
+                #tokens
+
+                use diesel::sqlite::{Sqlite, SqliteConnection};
+                use diesel::serialize::ToSql;
+                use diesel::deserialize::Queryable;
+
+                #[allow(dead_code)]
+                /// Registers an implementation for this function on the given connection
```

Comment:
Any chance to replace `this function` with the name of the function?

Reply:
It's possible, but would require more work than just `s/this function/#fn_name`. I'm actually not sure if it's even worth having these doc comments at all. The module they're being generated for is `#[doc(hidden)]` and if it wasn't, it's still only `pub(crate)` so this never gets rendered in docs. It's pretty much only there if you look at the expanded code